torch/csrc/autograd/python_function.cpp

```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   Py_INCREF((PyObject*)self);
   return std::shared_ptr<PyFunction>(&self->cdata, Decref());
 }
+
+namespace torch { namespace autograd {
+
+using memo_map = std::unordered_map<Node*, variable_list>;
+
+variable_list interpret_node(std::shared_ptr<Node>, input_map&, memo_map&);
+
+std::shared_ptr<Variable> interpret_output(Output output, input_map& inputs, memo_map& memo) {
+    auto vs = interpret_node(output.node, inputs, memo);
+    return vs.at(output.output_nr);
+}
+
+variable_list interpret_node(std::shared_ptr<Node> node, input_map& inputs, memo_map& memo) {
+    variable_list input_vars;
+    auto num_args = node->inputs.size();
+    input_vars.reserve(num_args);
+    for (auto input : node->inputs) {
+        input_vars.emplace_back(interpret_output(input, inputs, memo));
+    }
+    if (auto n = dynamic_cast<InputNode*>(node.get())) {
+        return {inputs.at(n)};
+    } else if (auto n = dynamic_cast<PyNode*>(node.get())) {
+        auto& cls = n->pyobj;
+        // Massage variables into form where we can THPFunction_apply it.
+        // While in principle we can avoid putting things into Python
+        // and then taking them out again, doing so seems to require an excess
+        // of faffing about to optimize a codepath that is already going to
+        // fundamentally be slow (since it calls into Python.) NOT. WORTH. IT.
+
+        PyObject* input_objs = PyTuple_New(num_args);
+        for (size_t i = 0; i < num_args; i++) {
+          PyTuple_SET_ITEM(input_objs, i, THPVariable_Wrap(input_vars.at(i)));
+        }
```

Comment:
Yes. The rewrite handles this correctly, but @zdevito convinced me that I should store the arguments out of band, and reconstruct the correct Python argument positions on the fly.